torch/functional.py

```diff
@@ -543,3 +544,241 @@ def argsort(input, dim=None, descending=False):
+            ord    matrix norm                   vector norm
+            =====  ============================  ==========================
+            None   Frobenius norm                2-norm
+            'fro'  Frobenius norm                --
+            'nuc'  nuclear norm                  --
+            inf    max(sum(abs(x), dim=1))       max(abs(x))
+            -inf   min(sum(abs(x), dim=1))       min(abs(x))
+            0      --                            sum(x != 0)
+            1      max(sum(abs(x), dim=0))       as below
+            -1     min(sum(abs(x), dim=0))       as below
+            2      largest singular value        as below
+            -2     smallest singular value       as below
+            other  as vec norm when dim is None  sum(abs(x)**ord)**(1./ord)
+            =====  ============================  ==========================
+        dim ({int, 2-tuple of ints, 2-list of ints}): If it is a int, vector norm
+        will be calculated, if it is 2-tuple of ints, matrix norm will be
+        calculated.
+        keepdim (bool): whether the output tensors have :attr:`dim`
+            retained or not. Ignored if ``dim=None``.
+        out (Tensor, optional) – the output tensor
+
+    Example::
+        >>> import torch
+        >>> a = torch.arange(9, dtype= torch.float) - 4
+        >>> b = a.reshape((3, 3))
+        >>> torch.norm(a)
+        tensor(7.7460)
+        >>> torch.norm(b)
+        tensor(7.7460)
+        >>> torch.norm(a, float('inf'))
+        tensor(4.)
+        >>> torch.norm(b, float('inf'))
+        tensor(9.)
+        >>> torch.norm(a, float('-inf'))
+        tensor(0.)
+        >>> torch.norm(b, float('-inf'))
+        tensor(2.)
+        >>> torch.norm(a, 1)
+        tensor(20.)
+        >>> torch.norm(b, 1)
+        tensor(20.)
+        >>> torch.norm(a, -1)
+        tensor(0.)
+        >>> torch.norm(b, -1)
+        tensor(0.)
+        >>> torch.norm(a, 2)
+        tensor(7.7460)
+        >>> torch.norm(b, 2)
+        tensor(7.7460)
+        >>> torch.norm(a, -2)
+        tensor(0.)
+        >>> torch.norm(b, -2)
+        tensor(0.)
+        >>> torch.norm(a, 3)
+        tensor(5.8480)
+        >>> torch.norm(a, -3)
+        tensor(0.)
+        >>> c = torch.tensor([[ 1, 2, 3],[-1, 1, 4]] , dtype= torch.float)
+        >>> torch.norm(c, dim=0)
+        tensor([1.4142, 2.2361, 5.0000])
+        >>> torch.norm(c, dim=1)
+        tensor([3.7417, 4.2426])
+        >>> torch.norm(c, p=1, dim=1)
+        tensor([6., 6.])
+        >>> d = torch.arange(8, dtype= torch.float).reshape(2,2,2)
+        >>> torch.norm(d, dim=(1,2))
+        tensor([ 3.7417, 11.2250])
+        >>> torch.norm(d[0, :, :]), torch.norm(d[1, :, :])
+        (tensor(3.7417), tensor(11.2250))
+    """
+    ndim = input.dim()
+
+    # catch default case
+    if dim is None and out is None:
+        if p == "fro":
+            return torch._C._VariableFunctions.norm(input, 2)
```

Comment:
Hmm this is calling the wrong function, right?

Reply:
It is intended. This follows the default case catch written in Numpy.norm. It also let current function works as the original torch.norm() considering backward compatibility.

Reply:
Sorry but I still don't quite understand here. Could you explain a bit more on why it is written this way on how it preserves BC?

Reply:
On production, we have a function "torch.norm(input, p=2) → Tensor". Therefore, there are numerous places in the codebase where it called .norm() directly. "fro" is the default value for attr:p now as Numpy did and "fro" is equal 2-norm when calling vector norm as I mentioned in the intro. With this condition check, .norm() will have the same behavior as before.

Reply:
I see. But ideally, if you call `torch._C._VariableFunctions.frobenius_norm` with `dim` being `None`, it should do the same thing as 2-norm in c++, right?